Minimal L1 Report Builder

NoahCarnahan opened the pull request.

NoahCarnahan committed:
[WIP] first steps adding report builder view

NoahCarnahan committed:
[WIP] fixed the sidebar issue..

Commit:
WIP cleaning up code

Commit:
remove view func and any references

Commit:
put 'Create Reports' behind toggle

Commit:
add real view

Commit:
add create new report form class

Commit:
pass create new report builder form in context

Commit:
render create new report builder form

Commit:
clean up view

Commit:
Choose application

Commit:
add support for forms in create new report

Commit:
redirects to next form

Commit:
refactor templates

Commit:
refactor views

Commit:
include kwarg for source_type

Commit:
pass kwargs to configure bar chart form

Commit:
add application to kwargs

Commit:
pass application to form

Commit:
populate group_by for cases

NoahCarnahan committed:
Fix import

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into report-builder

NoahCarnahan committed:
Merge branch 'generalize-case-property-template' into report-builder

NoahCarnahan committed:
Add sort property list to graph configuration page

NoahCarnahan committed:
Send sort property configuration in the form.

NoahCarnahan committed:
Use a select2 widget for the filter properties instead of a typeahead.

Commit:
adding structure for report builder post

Commit:
WIP generating data source and report

NoahCarnahan committed:
WIP fixed url reversal and data source saving bugs

NoahCarnahan committed:
Merge branch 'master' into report-builder

Conflicts:
	corehq/apps/app_manager/templates/app_manager/partials/case_list_properties.html
	corehq/apps/hqwebapp/models.py

NoahCarnahan committed:
Add columns to report

NoahCarnahan committed:
Rework report_filter_configuration.html template

No longer extend case_list_properties. Extending made things more
complicated, and we gained little. This makes both templates cleaner.

NoahCarnahan committed:
Merge branch 'master' into report-builder

Conflicts:
	corehq/apps/app_manager/templates/app_manager/partials/case_list_properties.html

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into report-builder

Conflicts:
	corehq/apps/app_manager/templates/app_manager/partials/case_list_properties.html

NoahCarnahan committed:
Removed forgotten merge conflict indicator

NoahCarnahan committed:
Remove reference to recently eliminated css file

NoahCarnahan committed:
Remove unused view model properties

NoahCarnahan committed:
Build data source

NoahCarnahan committed:
Renamed form field

NoahCarnahan committed:
Generate report filters

NoahCarnahan committed:
Configure bar chart

NoahCarnahan committed:
Limit graph reports to 10 filters maximum

NoahCarnahan committed:
Rename case_type_map to sources_map

NoahCarnahan committed:
Readable form names in report source select

NoahCarnahan committed:
[WIP] Sketch out outline for table report builder

NoahCarnahan committed:
Add missing file

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into report-builder

NoahCarnahan committed:
Allow for numeric type filters

NoahCarnahan committed:
[WIP] Sketch out form for table configuration

NoahCarnahan committed:
[WIP] Sketch out table configuration form

NoahCarnahan committed:
[WIP] Use inheritence in the report builder crispy forms

NoahCarnahan committed:
[WIP] Inherit template

NoahCarnahan committed:
[WIP] Don't create indicators for every case property

NoahCarnahan committed:
[WIP] Create form and views for pie chart report configuration

NoahCarnahan committed:
[WIP] update TODOs

NoahCarnahan committed:
[WIP] generate reports from table report configuration

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into report-builder

NoahCarnahan committed:
[WIP] fix typo

NoahCarnahan committed:
[WIP] Fix incorrect doc type on data source. Also give data sources unique table names

NoahCarnahan committed:
Rename function

NoahCarnahan committed:
Rearanged files and renamed classes

NoahCarnahan committed:
Reduce code reuse by not duplicating make indicator function

NoahCarnahan committed:
Share remaining indicator creation functions

NoahCarnahan committed:
Populate report configuration table with form questions.

This allows users to select form questions (and form metadata) for report columns and filters.
Still need to rework the report creation code to account for form data sources  though (currently
 only handles cases).

NoahCarnahan committed:
Use the filter field, not the slug, to reference table

NoahCarnahan committed:
[WIP] Add more support for creating reports from Forms

NoahCarnahan committed:
Don't start with an empty row in the filter and columns tables.

NoahCarnahan committed:
Add ability to filter cases by open/close status

NoahCarnahan committed:
[WIP] Escape question ids for use as indicator ids

NoahCarnahan committed:
Fix bug with select2 knockout binding

NoahCarnahan committed:
Hide 'Format' column of the report column configuration

NoahCarnahan committed:
Fix bug with form filters

Solved by removing "--" from the column name (and replaced with "__").
"--" seems to be legal in sql columns but there is an issue in how
these names get converted to sqlalchemy bind parameters. Looks like
the Select.append_whereclause() mehtod that we are using expects
bindparameters to only include [A-Za-z0-9_].

either an issue with
sqlagg or sqlalchemy in how these column names get escaped or turned into
queries.

NoahCarnahan committed:
Reorganize knockout bindings so that form validation can take place on view model.

NoahCarnahan committed:
Add validation in column and filter configuration tables

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into report-builder

Conflicts:
	corehq/apps/app_manager/templates/app_manager/partials/case_list_properties.html
	corehq/apps/userreports/views.py

NoahCarnahan committed:
[WIP] Reuse data sources

Create every possible indicator instead of just the ones needed for this report.

NoahCarnahan committed:
[WIP] Reuse data sources

Check for existing data sources that match the data source that
otherwise would be created exactly. This is fairly naive, and also
does a fair amount of docuemnt filtering in the django layer.
Writing a new couch view would speed this up.

NoahCarnahan committed:
Reuse data sources

Delete data sources that are no longer being used when deleting
reports. Again, the search here is done at the Django level. We
will likely want to move this into a new Couch view.

NoahCarnahan committed:
Show warning if data source hasn't been built yet

When viewing user configured reports, we now periodically check to see
if the data source has been built. After it has been built, automatically
refresh the report to show the new data. Current limitations:
- Doesn't migrate old data sources to add the built flag
- Changes existing report behavior: Whenever you view a UCR, the filters
  will be applied automatically when you load the page.
  (this wouldn't be too hard to prevent)
- Automatic application of current filters when data source is finished
  building could be jarring if the user starts investigating the data
  before it's all there.

NoahCarnahan committed:
Fix bug preventing some reports from being deleted.

NoahCarnahan committed:
First stab at limiting number of data sources

Enforces the limit in a pretty basic way. Still allowed to configure
unlimited data sources manually, but form builder created data sources
will not be limited. Check for "matching" data sources is inconsistent
with the way the code that actually creates data sources checks for
matches. These should probably use the same logic.

NoahCarnahan committed:
Improve report configuration form validation

NoahCarnahan committed:
Clean up notes

NoahCarnahan committed:
Fix bug with form validation

NoahCarnahan committed:
Share data source existence checking logic between forms.

The logic used to figure out if a suitable data source already exists
for a report was being repeated. This pulls that logic out into its
own class.

NoahCarnahan committed:
Add some doc strings

NoahCarnahan committed:
Use plus icon in sidebar

dimagimon commented:
There are 19 errors:

* corehq/apps/hqwebapp/models.py, line 271 - E124 closing bracket does not match visual indentation
* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/reports/builder/__init__.py, line 28 - E122 continuation line missing indentation or outdented
* corehq/apps/userreports/reports/builder/forms.py, line 304 - E501 line too long (152 > 115 characters)
* corehq/apps/userreports/reports/builder/forms.py, line 337 - E121 continuation line under-indented for hanging indent
* corehq/apps/userreports/reports/builder/forms.py, line 376 - E501 line too long (116 > 115 characters)
* corehq/apps/userreports/reports/builder/forms.py, line 473 - E303 too many blank lines (2)
* corehq/apps/userreports/reports/builder/forms.py, line 550 - E501 line too long (120 > 115 characters)
* corehq/apps/userreports/tests/test_app_manager_integration.py, line 62 - E303 too many blank lines (2)
* corehq/apps/userreports/views.py, line 14 - F401 'View' imported but unused
* corehq/apps/userreports/views.py, line 16 - F401 'Application' imported but unused
* corehq/apps/userreports/views.py, line 17 - F401 'ParentCasePropertyBuilder' imported but unused
* corehq/apps/userreports/views.py, line 34 - F401 'DEFAULT_CASE_PROPERTY_DATATYPES' imported but unused
* corehq/apps/userreports/views.py, line 94 - E501 line too long (118 > 115 characters)
* corehq/apps/userreports/views.py, line 401 - E302 expected 2 blank lines, found 1
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 675 - ['value'] is better written in dot notation.
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 637 - Weird construction. Is 'new' necessary?

czue commented on corehq/apps/userreports/reports/builder/forms.py:
```diff
@@ -0,0 +1,569 @@
+            crispy.Fieldset(
+                _('Create New Report'),
+                'report_type',
+                crispy.Field('application', data_bind='value: application'),
```
probably worth cross-linking this and the js with comments

NoahCarnahan replied:
Done. Thanks.

czue commented on corehq/apps/userreports/reports/builder/forms.py:
```diff
@@ -0,0 +1,569 @@
+                crispy.Field('report_source', data_bind='''
+                    options: sourcesMap[application()][sourceType()],
+                    optionsText: function(item){return item.text},
+                    optionsValue: function(item){return item.value}
```
would be awesome to reuse this functionality in the "create data source from form/case" UI

NoahCarnahan replied:
Added in 84bee6d. Unfortunately I didn't come up with a great way to encapsulate this functionality.

czue replied:
thoughts on this attempt? https://github.com/dimagi/commcare-hq/pull/5793

czue commented on corehq/apps/userreports/reports/builder/forms.py:
```diff
@@ -0,0 +1,569 @@
+                    if (
+                        s.referenced_doc_type == ds_builder.source_doc_type and
+                        s.configured_filter == ds_builder.filter and
+                        s.configured_indicators == ds_builder.indicators
```
probably easier to add explicit references to app/form/case here and timestamp them or link to a build id

NoahCarnahan replied:
Updated in 5298395 to use app + xmlns/case type + app build number. There's a potentially big downside here though in that any app change will "invalidate" the data source, even if it doesn't affect the case type or form that the data source is based on. But, perhaps this is satisfactory until a more comprehensive solution is found (such as calculating diffs between needed data sources and existing data sources or something like that).

czue commented on corehq/apps/userreports/reports/builder/forms.py:
```diff
@@ -0,0 +1,569 @@
+
+        def escape_id(id):
+            # TODO: This is fairly naive escaping
+            return id.strip("/").replace("/", "__")
```
can use something like this: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/sql.py#L80

NoahCarnahan replied:
Fixed in 2b1fef6, thanks!

snopoke commented on corehq/apps/userreports/views.py:
```diff
@@ -80,6 +202,26 @@ def _edit_report_shared(request, domain, config):
+    # Delete the data source too if it's not being used by any other reports.
+    # If we decide to go this route, we most likely should write a new couch
+    # view to quickly get the ReportConfigurations that reference this data
+    # source.
```
New couch view seems like a good idea.

czue replied:
+1

NoahCarnahan replied:
Fixed with b2cd306, thanks.

snopoke commented on corehq/apps/userreports/tasks.py:
```diff
@@ -31,6 +31,8 @@ def rebuild_indicators(indicator_config_id):
             logging.exception('problem saving document {} to table. {}'.format(doc['_id'], e))
     adapter.engine.dispose()
+    config.built = True
+    config.save()
```
Could probably use `django-soil` here to keep track of progress and completion status.

NoahCarnahan replied:
I didn't know about `django-soil`, thanks for pointing that out.
As currently written, I check if the data source has been built every time the report is
loaded. That means I'll probably be checking the built status after the `soil.DownloadBase`
is no longer in the cache. Does soil have a good way to handle this situation?

I could still store the built status on the `DataSourceConfiguration`. Then, when the report
is loaded I would first look to get the status from the `DownloadBase`, and if it doesn't
exist, look on the `DataSourceConfiguration` itself. Seems sorta redundant though...

snopoke replied:
I was thinking more of being able to give more of a progress dialog to the user. You could still use this to determine if the config has been built but I would suggest setting `built = False` when changes get made to the config.

NoahCarnahan replied:
Makes sense, thanks.

czue commented on corehq/apps/userreports/reports/builder/forms.py:
```diff
@@ -0,0 +1,569 @@
+
+    @property
+    def column_config_template(self):
+        path = os.path.join(
```
i think you can use the django template loader for this

NoahCarnahan replied:
Implemented in cc1f5f5, thanks for the tip.

snopoke commented:
Did a quick pass and looks good. Do we have an idea of how to handle app changes when they happen?
Does the user just come back and update the report config?

czue commented on corehq/apps/userreports/reports/builder/forms.py:
```diff
@@ -0,0 +1,569 @@
+
+        # Check if a suitable data source already exists.
+        # This checking is pretty naive. Do something better.
+        # Also inefficient. Consider writing a new couch view.
```
+1 - same comment as above

NoahCarnahan commented:
Yeah that's right @snopoke, if you make any app changes that affect the report you need to edit the config "by hand" or use the report build to create the report again. (Basically we're just push that problem out of scope for now :) )

NoahCarnahan committed:
Fix lint errors

dimagimon commented:
There are 5 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/views.py, line 14 - F401 'View' imported but unused
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 637 - Weird construction. Is 'new' necessary?

NoahCarnahan committed:
resolve lint error

dimagimon commented on corehq/apps/reports/urls.py:
```diff
@@ -19,6 +25,10 @@
     TestReport.url_pattern(),
     ConfigurableReport.url_pattern(),
+    url(r'^builder/create/$', CreateNewReportBuilderView.as_view(), name="create_new_report_builder"),
+    url(r'^builder/configure/bar/$', ConfigureBarChartReportBuilderView.as_view(), name="configure_bar_chart_report_builder"),
```
E501 line too long (126 > 115 characters)

dimagimon commented on corehq/apps/reports/urls.py:
```diff
@@ -19,6 +25,10 @@
     ConfigurableReport.url_pattern(),
+    url(r'^builder/create/$', CreateNewReportBuilderView.as_view(), name="create_new_report_builder"),
+    url(r'^builder/configure/bar/$', ConfigureBarChartReportBuilderView.as_view(), name="configure_bar_chart_report_builder"),
+    url(r'^builder/configure/pie/$', ConfigurePieChartReportBuilderView.as_view(), name="configure_pie_chart_report_builder"),
```
E501 line too long (126 > 115 characters)

dimagimon commented on corehq/apps/reports/urls.py:
```diff
@@ -19,6 +25,10 @@
+    url(r'^builder/create/$', CreateNewReportBuilderView.as_view(), name="create_new_report_builder"),
+    url(r'^builder/configure/bar/$', ConfigureBarChartReportBuilderView.as_view(), name="configure_bar_chart_report_builder"),
+    url(r'^builder/configure/pie/$', ConfigurePieChartReportBuilderView.as_view(), name="configure_pie_chart_report_builder"),
+    url(r'^builder/configure/table/$', ConfigureTableReportBuilderView.as_view(), name="configure_table_report_builder"),
```
E501 line too long (121 > 115 characters)

dimagimon commented on corehq/apps/style/static/style/ko/knockout_bindings.ko.js:
```diff
@@ -629,6 +629,53 @@ ko.bindingHandlers.typeahead = {
+ * a list of strings, or a list of objects with the keys 'id' and 'text' used
+ * for the select2's options.
+ */
+ko.bindingHandlers.select2 = new function(){
```
Weird construction. Is 'new' necessary?

czue replied:
is @dimagimon becoming sentient?

snopoke replied:
Was also a bit puzzled by this.

NoahCarnahan replied:
I wanted to create an object where both the `init` and `update` properties refer to the `SOURCE_KEY` property, so I couldn't use an object literal. I think the anonymous constructor approach that I used looks a little cleaner than either breaking the object's definition over multiple statements or using a "self-calling" anonymous function. Is there a best practice or style guideline for this situation?
(I found a somewhat related discussion on [SO](http://stackoverflow.com/questions/1536127/module-pattern-vs-instance-of-an-anonymous-constructor))

NoahCarnahan committed:
Use Django template loader instead of reading file manually

NoahCarnahan committed:
Add case type/form filtering to selects in 'Form/Case data source from app' forms

NoahCarnahan committed:
Ensure that data source column names are unique

dimagimon commented:
There are 1 errors:

* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file

NoahCarnahan committed:
Add couch view for look up of report configurations by data source id

NoahCarnahan committed:
Fix bug where users cannot leave filters blank

dimagimon commented:
There are 2 errors:

* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan committed:
Add notes

NoahCarnahan committed:
Use couch view and new meta data to find suitable DataSourceConfigurations

NoahCarnahan committed:
[WIP] Update what is saved in DataSourceConfiguration meta field

There's a bug with the "edit data source" form that I may have introduced.
Trying to get to the bottom of it...

dimagimon commented:
There are 6 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan commented:
Thanks for the feedback guys. I've the addressed the comments in the latest commits. Unfortunately it looks like I broke the non report builder "Edit data source" form though, so that needs to be fixed prior to merging. Otherwise, I think this is ready and I believe the plan is to not go through QA prior to merging since this is under a flag. Amelia has also suggested some UI changes that I'll be working on in ~2 weeks when my interrupt duty is over.

NoahCarnahan commented:
@czue @snopoke 
I tracked down the bug I was chasing, but I'm not sure what the cleanest way to fix it is.
This is what happened:
- Some bad commits were introduced on master. It was one of these:
    029d2ea957a837c11801b0c4a10bf6d32fbb8f98
    be782c5774d6705a529cf67762b307bfe0efb0da
    bfd9392d0d552e98d5e7c4f5ab50cdf7fdedc71a
- I merged those commits into my branch when I merged master into my branch a while ago.
- At a later point, the bug was fixed by 8e3073e6ba001aa85c9e36b86aa401fcc588eeeb
- The bug is fixed on master, but if I merge master into my branch the bug still remains on my branch.

I'm not sure what the cleanest way to fix this is. I tried cherry-picking the commit that fixes it, but that does nothing. Should I rebase my branch off of master? I generally try to avoid rewriting history though. I suppose I could manually make the edits introduced by the fixing commit onto my branch.
Thoughts? I'm not sure why merging master in the first place didn't fix things.

(For reference, the bug is that saving changes on the data source configuration form has no effect - form is interpreted as invalid)

czue commented:
@NoahCarnahan can you just read https://github.com/dimagi/commcare-hq/commit/8e3073e6ba001aa85c9e36b86aa401fcc588eeeb and figure out if there was a bad merge that caused not the whole fix to be applied?

The main issue was that the form arguments being passed in and up the inheritance chain were not consistent with what django's form classes expected so somewhere we were passing the wrong value for some arguments.

NoahCarnahan committed:
[WIP] Implemented changes found in 8e3073e

NoahCarnahan committed:
Fix data source form bugs

NoahCarnahan committed:
Remove unused method

dimagimon commented:
There are 6 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan committed:
Revert "Remove unused method"

This reverts commit 047fe1740a3d35502fbf0dbe451ea95573e9c0f1.

NoahCarnahan committed:
Remove note

dimagimon commented:
There are 6 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan commented:
@czue 
manually creating a commit that replicates 8e3073e did the trick.
In the course of debugging this I found a different bug though (form questions with no label don't show up in select2), so let me fix that before we merge.

NoahCarnahan commented:
(False alarm on the other bug)

czue commented:
@NoahCarnahan are you ready for "final" review now?

NoahCarnahan commented:
Yup, I think so.

On Mon, Mar 2, 2015 at 4:13 PM, Cory Zue <notifications@github.com> wrote:

> @NoahCarnahan <https://github.com/NoahCarnahan> are you ready for "final"
> review now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5608#issuecomment-76717842>.
>

czue commented:
@NoahCarnahan a few small comments, but I'm pretty much signed off. where did we land on putting this into a subscription? the only reason to do it now is so that we don't forget that we enabled it for domains that weren't on the right subscription now (which I think isn't a bad reason). otherwise when we do add that restriction we'll break them, right?

NoahCarnahan committed:
Add missing space

NoahCarnahan committed:
Check for existence of attributes in couch view

NoahCarnahan committed:
Remove redundant data_sources_by_domain couch view

NoahCarnahan committed:
Pull matching data source lookup into function

dimagimon commented:
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/models.py, line 208 - E501 line too long (116 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan commented:
@czue I forgot about the subscription question. I wasn't really involved with that discussion, but @amsagoff might know where that stands? (feel free to take to email if more appropriate). Right now this is under the general "User Configurable Reports UI" flag. I think it's probably best to implement the subscription constraint now.

czue commented:
@NoahCarnahan i think for now we can make it pro or higher. you can refer to @nickpell's recent PR for a template of how to do it: https://github.com/dimagi/commcare-hq/pull/5750

NoahCarnahan commented:
Great, will do.

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into report-builder

czue committed:
refactor the app source into a reusable field that can be dropped in forms

update the report builder UI and UCR data source from apps to use this

czue commented:
also i think the data source lookup by domain might not be filtering properly. i see all data sources in the sidebar on every domain

czue committed:
convert js to anonymous functions

czue committed:
better docs

czue committed:
don't use the django MultiField framework for app sources

just rely on a bit more scaffolding to tie things together

NoahCarnahan committed:
Add missing end_key to couch view query

dimagimon commented:
There are 9 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/models.py, line 208 - E501 line too long (116 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/views.py, line 343 - E303 too many blank lines (3)
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 637 - Weird construction. Is 'new' necessary?
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan commented:
@czue, just fixed the source lookup by domain. I had forgotten the `end_key`.

czue committed:
Merge pull request #5793 from dimagi/report-builder-cz

refactor the app source into a reusable field that can be dropped in forms

dimagimon commented:
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 103 - W391 blank line at end of file
* corehq/apps/userreports/ui/forms.py, line 11 - F401 'memoized' imported but unused
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan committed:
Add report builder to the pro plan and higher

NoahCarnahan committed:
Merge branch 'report-builder' of github.com:dimagi/commcare-hq into report-builder

Conflicts:
	corehq/apps/userreports/views.py

dimagimon commented:
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 103 - W391 blank line at end of file
* corehq/apps/userreports/ui/forms.py, line 11 - F401 'memoized' imported but unused
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan commented:
@czue I've put the report builder under the pro plan (see  b990e9f). It requires the running of a management command `./manage.py cchq_prbac_bootstrap`. Is it better with regard to our deployment process if I put this in a migration like Nick did here: a0057a9425abdc77ef906341038d601fc5edac65 ?

nickpell commented:
+1 automating deploys

NoahCarnahan committed:
bootstrap prbac in a migration

dimagimon commented:
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 103 - W391 blank line at end of file
* corehq/apps/userreports/ui/forms.py, line 11 - F401 'memoized' imported but unused
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

NoahCarnahan commented:
management command will now run in a migration.

NoahCarnahan commented:
I think this beast is ready for merge.

czue commented:
:eyes:
